src/Symfony/Component/Messenger/Middleware/LockMiddleware.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Message\LockableMessage;
+use Symfony\Component\Messenger\Stamp\LockStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+
+final class LockMiddleware implements MiddlewareInterface
+{
+    private ?LockFactory $lockFactory = null;
+
+    public function __construct(
+        ?LockFactory $lockFactory,
+    ) {
+        $this->lockFactory = $lockFactory;
+    }
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        if (null === $this->lockFactory) {
+            return $stack->next()->handle($envelope, $stack);
+        }
+
+        $message = $envelope->getMessage();
+
+        // If we're trying to dispatch a lockable message.
+        if ($message instanceof LockableMessage && null === $envelope->last(ReceivedStamp::class)) {
+            $key = $message->getKey();
+
+            if (null !== $key) {
+                // The acquire call must be done before stamping the message
+                // in order to have the full state of the key in the stamp.
+                $canAcquire = $this->lockFactory->createLockFromKey($key, autoRelease: false)->acquire();
```

Comment:
Is there a mechanism that asserts the key can be serialized?
If people created a lockFactory with stores like flock, I'm pretty sure it would not work as expected (The key contains a file resource and won't be serialized)

Reply:
Indeed, seems like store like FlockStore is an issue since they call `Key::markUnserializable`.
But like some other store, if we don't have the state of the key we cannot release it
https://github.com/symfony/symfony/blob/3fdc2c99d101ad69dd5c99ed03d8d64ad9608631/src/Symfony/Component/Lock/Store/FlockStore.php#L131-L133

Do you see a solution for this ? Or should this middleware be limited to serializable key ?

Reply:
maybe their is nothing to do, and just let people deal with the error `UnserializableKeyException`.

If we want to be smarter, we would need to add an interface marker on Store and provide a method on the factory (or lock generated by the factory), but I'm not sure it worth it